api/api/src/main/scala/uk/ac/wellcome/platform/api/rest/QueryParams.scala

```diff
@@ -48,23 +48,38 @@ trait QueryParamsUtils extends Directives {
   def decodeOneWithDefaultOf[T](default: T, values: (String, T)*): Decoder[T] =
     Decoder.decodeString.map { values.toMap.getOrElse(_, default) }
 
-  def decodeOneOfCommaSeparated[T](values: (String, T)*): Decoder[List[T]] =
+  def decodeOneOfCommaSeparated[T](values: (String, T)*): Decoder[List[T]] = {
+    val mapping = values.toMap
+    val validStrs = values.map(_._1).toList
     decodeCommaSeparated.emap { strs =>
-      val mapping = values.toMap
-      val results = strs.map { str =>
-        mapping
-          .get(str)
-          .map(Right(_))
-          .getOrElse(Left(str))
-      }
-      val invalid = results.collect { case Left(error) => error }
-      val valid = results.collect { case Right(value)  => value }
-      (invalid, valid) match {
-        case (Nil, results) => Right(results)
-        case (invalidValues, _) =>
-          Left(invalidValuesMsg(invalidValues, values.map(_._1).toList))
-      }
+      mapStringsToValues(strs, mapping).left
+        .map { invalidStrs =>
+          invalidValuesMsg(invalidStrs, validStrs)
+        }
     }
+  }
+
+  def decodeIncludesAndExcludes[T](
+    values: (String, T)*): Decoder[(List[T], List[T])] = {
+    val mapping = values.toMap
+    val validStrs = values.map(_._1).toList
+    decodeCommaSeparated
+      .emap { strs =>
+        val (excludeStrs, includeStrs) = strs.partition(_.startsWith("!"));
+        val includes = mapStringsToValues(includeStrs, mapping);
+        val excludes = mapStringsToValues(excludeStrs.map(_.tail), mapping);
```

Comment:
Not too fussed but would `_.stripSuffix("!")` be more clear than `.tail`?

Reply:
`tail` works for me, and means it's all on one line :see_no_evil:... but can do the `stripPrefix` if you think `tail` is confusing?

Reply:
`stripPrefix` will actually do unnecessary work (checking the string starts with `!` when we know that all of them do), but I don't suppose that matters

Reply:
Na, not really, I was probably just reading line by line rather than the whole method, whatever works for you.

Reply:
🚀 performance!